yt_idefix/data_structures.py

```diff
@@ -320,3 +321,98 @@ def _parse_parameter_file(self):
 class PlutoVtkDataset(IdefixVtkDataset):
     _version_regexp = re.compile(r"\d+\.\d+\.?\d*[-\w+]*")
     _required_header_keyword = "PLUTO"
+
+    def __init__(
+        self,
+        filename,
+        dataset_type=None,
+        unit_system="cgs",
+        units_override=None,
+        *,
+        code_dir: str | None = None,
+        geometry: str | None = None,
+        inifile=None,
+    ):
+        # Presuming code files are along with output, otherwise header file location
+        # has to be specified.
+        if code_dir is None:
+            self.code_dir = os.path.abspath(os.path.dirname(filename))
+        else:
+            self.code_dir = os.path.abspath(code_dir)
+
+        super().__init__(
+            filename,
+            dataset_type=dataset_type,
+            unit_system=unit_system,
+            units_override=units_override,
+            geometry=self.geometry,
+            inifile=inifile,
+        )
+
+    def _parse_parameter_file(self):
+        self._parse_header_file()
+        super()._parse_parameter_file()
+        self._get_time()
+
+    def _parse_header_file(self):
+        """Pluto reads some metadata from header file 'definitions.h'."""
+        constexpr = re.compile(r"CONST_\w*")
+        header_file = os.path.join(self.code_dir, "definitions.h")
+        if os.path.isfile(header_file):
+            with open(header_file) as hf:
+                for line in hf.readlines():
+                    if "GEOMETRY" in line:
+                        self.geometry = line.split()[-1].lower()
+                    if "UNIT" in line:
+                        item = line.split()
+                        key = item[1].lower()
+                        expr = item[-1]
+                        if "CONST" in expr:
+                            expr = constexpr.sub(self._get_const_for_reg, expr)
+                        self.parameters[key] = eval(expr)
```

Comment:
the implementation is rather confusing to me. I don't think I can maintain that without rewriting/simplifying it. I'm fine with refactoring it myself later but I'll let you the initiative if you feel like doing it.

In particular, it's a little unsettling that pluto_def_constants' values are floats, but you're only ever using them as `str`, this feels like it could be improved.

Reply:
I should have left more comments in this part. For the cases where the pre-defined constant is used to define code units, such as `1000*CONST_pc`, the expression is a string and I replace the constant with its value. Then the code unit is able to be calculated. 

This might be a stupid way but works. I'm wiiling to rewrite it if there's a better way.

> In particular, it's a little unsettling that pluto_def_constants' values are floats, but you're only ever using them as `str`, this feels like it could be improved.

The values stored as floats is intuitive although I use them as `str` later. I can stored them as `str` if necessary.

Reply:
the main problem I see with your implementation is that it may turn into unspecified behaviours with valid data

For instance
```python
for line in hf.readlines():
    if "GEOMETRY" in line:
        self.geometry = line.split()[-1].lower() 
```

would produce problematic results with the following content
```C
#define GEOMETRY CARTESIAN
// GEOMETRY could be set to POLAR
```
with your current implementation, we'd get `self.geometry = "polar"` while the correct result should be `"cartesian"`

We should parse it in a much more restricted fashion to avoid this bug, for instance
```python
geom_regexp = re.compile(r"#define\s+GEOMETRY\s+([A-Z]+)")
for line in lines:
    match = re.search(geom_regexp, line)
    if match is None:
        continue
    self.geometry = match.group(1).lower()
```

ignoring multi-lines C comments would also be useful but it's harder.

Reply:
I see. I'll rewrite it, thanks!

Reply:
btw, I see you used `hf` instead of `fh` (file handle) as variable name

Reply:
`hf` is short for header file. Is it for convension to use `fh`?

Reply:
In yt (and by extension yt_idefix), the dominant convention is `fh` for file handle whatever the file, I'd prefer to keep it consistent.

Reply:
I thought about the problems with the accidentally parsing comments and my solution isn't actually much more robust than yours. The difficulty comes from that we're parsing the header file line by line, whereas it would be much easier to do it correctly if we stripped out comments first.

Reply:
Inspiring by your example of geometry, now I use regular expression to identify the true expression. I think the comments won't be accidentally included.

Reply:
How about this ?
```
#define GEOMETRY CARTESIAN
//#define GEOMETRY POLAR
```
And it gets worse with multi-line comments.
Anyway, that's what tests are for :)

Reply:
I use `"^\s*#define\s+GEOMETRY\s+([A-Z]+)"` in the new commit. It can ignore any comments, I think.

Reply:
That'll do